volatility3/framework/plugins/windows/pe_symbols.py

```diff
@@ -0,0 +1,777 @@
+
+    cached_value = Union[int, str, None]
+    cached_value_dict = Dict[str, Union[Dict[str, List[str]], Dict[str, List[int]]]]
+
+    def __init__(
+        self,
+        layer_name: str,
+        mod_name: str,
+        module_start: int,
+        symbol_module: Union[interfaces.context.ModuleInterface, pefile.ExportDirData],
+    ):
+        self._layer_name = layer_name
+        self._mod_name = mod_name
+        self._module_start = module_start
+        self._symbol_module = symbol_module
+
+        self._address_cache: PESymbolFinder.cached_int_dict = {}
+        self._name_cache: PESymbolFinder.cached_str_dict = {}
+
+    def _get_cache_key(self, value: cached_value) -> str:
+        """
+        Maintain a cache for symbol lookups to avoid re-walking of PDB symbols or export tables
+        within the same module for the same address in the same layer
+        """
+        return f"{self._layer_name}|{self._mod_name}|{value}"
+
+    def get_name_for_address(self, address: int) -> cached_str:
+        cached_key = self._get_cache_key(address)
+        if cached_key not in self._name_cache:
+            name = self._do_get_name(address)
+            self._name_cache[cached_key] = name
+
+        return self._name_cache[cached_key]
+
+    def get_address_for_name(self, name: str) -> cached_int:
+        cached_key = self._get_cache_key(name)
+        if cached_key not in self._address_cache:
+            address = self._do_get_address(name)
+            self._address_cache[cached_key] = address
+
+        return self._address_cache[cached_key]
+
+    def _do_get_name(self, address: int) -> cached_str:
+        raise NotImplementedError("_do_get_name must be overwritten")
+
+    def _do_get_address(self, name: str) -> cached_int:
+        raise NotImplementedError("_do_get_address must be overwritten")
+
+
+class PDBSymbolFinder(PESymbolFinder):
+    """
+    PESymbolFinder implementation for  PDB modules
+    """
+
+    def _do_get_address(self, name: str) -> PESymbolFinder.cached_int:
+        try:
+            return self._symbol_module.get_absolute_symbol_address(name)
+        except exceptions.SymbolError:
+            return None
+
+    def _do_get_name(self, address: int) -> PESymbolFinder.cached_str:
+        try:
+            name = self._symbol_module.get_symbols_by_absolute_location(address)[0]
+            return name.split(constants.BANG)[1]
+        except (exceptions.SymbolError, IndexError):
+            return None
+
+
+class ExportSymbolFinder(PESymbolFinder):
+    """
+    PESymbolFinder implementation for  PDB modules
+    """
+
+    def _get_name(self, export: pefile.ExportData) -> Optional[str]:
+        # AttributeError throws on empty or ordinal-only exports
+        try:
+            return export.name.decode("ascii")
+        except AttributeError:
+            return None
+
+    def _do_get_name(self, address: int) -> PESymbolFinder.cached_str:
+        for export in self._symbol_module:
```

Comment:
Yes, but again it won't be due to the specific code paths per implementation.

Reply:
So, they have the same "interface" in the sense that they have a function with a similar name that happens to take two inputs (of different types) and returns the same value?  Wow, interfaces have let themselves slip since I last used them...  5;P